### PR TITLE
Available CRDs check feature

### PR DESCRIPTION
**Pull request by raaizik**
## Changes

- Adds a new controller that reacts whenever a create/delete event has been enqueued for a specific CRD that's become newly available. 
- Note: this PR introduces only the structure of the feature without any specific CRDs which will be added from their respective PRs (e.g., `ClusterClaim`, `VolumeGroupSnapshotClass` and `DRClusterConfig`).

### CODE REVIEW

**Review by iamniting — CHANGES_REQUESTED** [review]
/hold

Why do we need this PR? I really discourage having such changes without the proper discussion with the maintainers.

Also, I would highly encourage having meaningful commit title and message. The current one does not make any sense.

Screenshot from 2024-07-25 11-08-27

**Review by iamniting — CHANGES_REQUESTED** [review]
Could you please separate this commit into two commits?

* The first commit should include the controller changes.
* The second commit should include the script changes.

I will help test the first commit without the script. I believe we should be sorted with just the first commit.

**Review by umangachapagain — CHANGES_REQUESTED** [review]
This feels like a complex hack to cause a panic without actually using `panic()`.
How did we finalize on this approach?

**Comment by iamniting** [comment]
> This feels like a complex hack to cause a panic without actually using `panic()`. How did we finalize on this approach?

@umangachapagain We do not need this approach at all, Me and Rewant tested the solution without any script and fixed the bugs we had and it works fine. They will soon update the PR.

**Review by iamniting — CHANGES_REQUESTED** [review]
```
- At the start each reconcile iteration, revalidate which CRD are now
 available. If a CRD of interest is now avail, panic the op
```

Can you pls rephrase this statement in the commit message, This is not correct. Also please update the commit title.

File: controllers/util/k8sutil.go
Comment (iamniting) [inline controllers/util/k8sutil.go:48]: Can we rename this variable and please add a comment on what does it holds.

File: controllers/crd/crd_controller.go
Comment (iamniting) [inline controllers/crd/crd_controller.go:30]: I think only name is enough here.

**Review by iamniting — CHANGES_REQUESTED** [review]

File: controllers/util/k8sutil.go
Comment (raaizik) [reply controllers/util/k8sutil.go:48]: > Note: this PR introduces only the structure of the feature without any specific CRDs which will be added from their respective PRs (e.g., ClusterClaim, VolumeGroupSnapshotClass and DRClusterConfig).

Comment (iamniting) [reply controllers/util/k8sutil.go:48]: We also need this to be tested with an empty list. We should not assume that it will work.

Comment (rewantsoni) [reply controllers/util/k8sutil.go:48]: Tested with with and without entries in CRDList and it works

File: controllers/crd/crd_controller.go
Comment (nb-ohad) [inline controllers/crd/crd_controller.go:37]: I do not believe a map guarantees the order of keys if the 2 maps contain different keys. DeepEqual might fail even if both maps contain the exact same keys.  

Please move to a "len check + loop on items" for checking equality.

File: controllers/crd/crd_controller.go
Comment (nb-ohad) [inline controllers/crd/crd_controller.go:39]: I don't believe panic will allow us to detect the reason for the exit from outside the operator. use `os.Exit` with a specific exit code.

File: controllers/util/k8sutil.go
Comment (nb-ohad) [inline controllers/util/k8sutil.go:48]: Global singleton non-cost, non-atomic state inside a utils file is a red flag. Why do we need this?

**Review by nb-ohad — CHANGES_REQUESTED** [review]
Maybe I am missing something, but the current code will panic the entire pod! 
Why have we removed the code that recycles the process without exiting failing the pod?

**Comment by nb-ohad** [comment]
@iamniting @umangachapagain 
This approach it not the approach that was discussed. We need to talk about this before merging this one.

**Comment by nb-ohad** [comment]
/hold
Adding a hold until we have a proper discussion and agreement on the approach

File: controllers/storagecluster/reconcile.go
Comment (nb-ohad) [inline controllers/storagecluster/reconcile.go:173]: Available CRD is a map. I am not sure if there is an order guarantee on keys which means that DeepEqual might fail even if all the keys and values are the same. I would suggest changing the check to `maps.Equal`

File: controllers/storagecluster/storagecluster_controller.go
Comment (nb-ohad) [inline controllers/storagecluster/storagecluster_controller.go:1]: Where is the conditional watch that checks the CRD exists before adding the watch?

File: controllers/storagecluster/reconcile.go
Comment (nb-ohad) [inline controllers/storagecluster/reconcile.go:172]: Can we please have the exit code in a constant with a name that explains the usage? 
In addition, a comment about this line will help

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: hack/crdavail.sh
Comment (nb-ohad) [inline hack/crdavail.sh:11]: There is a missing newline here and git is complaining

File: hack/crdavail.sh
Comment (leelavg) [inline hack/crdavail.sh:6]: we are effectively making this a sub-process rather than a PID 1 and I'm not confident enough in the shell handling signals vs controller-runtime.

I could see good amount of comments on this PR, in brief, may I know why did we backoff from kubelet restarting us?

File: controllers/util/k8sutil.go
Comment (leelavg) [inline controllers/util/k8sutil.go:49]: is my understanding correct that in upcoming PRs we register the CRDNames in this and conditionally watch on them in the setupwithmanager?

File: controllers/storagecluster/storagecluster_controller.go
Comment (leelavg) [inline controllers/storagecluster/storagecluster_controller.go:231]: is there a possibility that CRD deletion is less than our operator lifetime?

File: controllers/storagecluster/storagecluster_controller.go
Comment (leelavg) [inline controllers/storagecluster/storagecluster_controller.go:257]: we already made a mistake (with Virt) in watching all CRDs increasing the memory footprint, my mistake in not pushing hard on this fix https://github.com/red-hat-storage/ocs-operator/pull/2539, seeing this PR we don't want to watch all CRDs until & unless necessary.

Clubbed w/ 2539 we could do something like below as we are only interested in the metadata of CRDs while mapping CRDs.
```
crd := &metav1.PartialObjectMetadata{}
crd.SetGroupVersionKind(extv1.SchemeGroupVersion.WithKind("CustomResourceDefinition"))
crd.Name = <CRDName>
client.Get(<ctx>, <ns-name>, crd)
```

**Review by leelavg — COMMENTED** [review]
I believe it might've been discussed, just out of curiosity, why can't the operation be offloaded to odf-op? odf-op could set env and ocs-op correspondingly sets up the watch? I didn't think of all the possibilities though.

File: controllers/storagecluster/reconcile.go
Comment (rewantsoni) [reply controllers/storagecluster/reconcile.go:173]: I tried to create a map, added keys in a different order, and compared them using DeepEqual and it worked. maps.Equal produces same results
https://go.dev/play/p/uFkmp57xEIp

File: controllers/storagecluster/storagecluster_controller.go
Comment (rewantsoni) [reply controllers/storagecluster/storagecluster_controller.go:1]: Added it in https://github.com/red-hat-storage/ocs-operator/pull/2745

File: controllers/storagecluster/reconcile.go
Comment (rewantsoni) [reply controllers/storagecluster/reconcile.go:172]: Added it with https://github.com/red-hat-storage/ocs-operator/pull/2745

File: hack/crdavail.sh
Comment (rewantsoni) [reply hack/crdavail.sh:11]: Done

File: controllers/util/k8sutil.go
Comment (rewantsoni) [reply controllers/util/k8sutil.go:49]: Added it with https://github.com/red-hat-storage/ocs-operator/pull/2745

File: controllers/storagecluster/storagecluster_controller.go
Comment (rewantsoni) [reply controllers/storagecluster/storagecluster_controller.go:231]: Although it is very unlikely but it is possible, and if they happens it will lead to a pod restart

File: controllers/storagecluster/storagecluster_controller.go
Comment (rewantsoni) [reply controllers/storagecluster/storagecluster_controller.go:257]: Looked at this and I couldn't find a way to specify multiple names in the cache using field selector, But I have updated it to use PartialObjectMetadata here: https://github.com/red-hat-storage/ocs-operator/pull/2745

**Pull request closed without merging**
